x/storage/keeper/payment.go

```diff
@@ -247,81 +247,23 @@ func (k Keeper) UnChargeObjectStoreFee(ctx sdk.Context, primarySpId uint32, buck
-		return fmt.Errorf("get GVG family failed: %d", bucketInfo.GlobalVirtualGroupFamilyId)
-	}
-
-	paymentAddr := sdk.MustAccAddressFromHex(bucketInfo.PaymentAddress)
-	price, err := k.spKeeper.GetGlobalSpStorePriceByTime(ctx, internalBucketInfo.PriceTime)
-	if err != nil {
-		return fmt.Errorf("get storage price failed: %d %w", internalBucketInfo.PriceTime, err)
-	}
-
-	// primary sp total rate
-	primaryTotalFlowRate := sdk.ZeroInt()
-
-	primaryRate := price.PrimaryStorePrice.MulInt(sdkmath.NewIntFromUint64(chargeSize)).TruncateInt()
-	if primaryRate.IsPositive() {
-		primaryTotalFlowRate = primaryRate
-		_, err = k.paymentKeeper.UpdateStreamRecordByAddr(ctx, types.NewDefaultStreamRecordChangeWithAddr(sdk.MustAccAddressFromHex(gvgFamily.VirtualPaymentAddress)).
-			WithStaticBalanceChange(primaryTotalFlowRate.MulRaw(timeToPay)))
-		if err != nil {
-			return fmt.Errorf("pay GVG family failed: %s %s %s", bucketInfo.BucketName, objectInfo.ObjectName, err)
-		}
-	}
-
-	// secondary sp total rate
-	secondaryTotalFlowRate := sdk.ZeroInt()
-
-	var lvg *storagetypes.LocalVirtualGroup
-	for _, l := range internalBucketInfo.LocalVirtualGroups {
-		if l.Id == objectInfo.LocalVirtualGroupId {
-			lvg = l
-			break
-		}
-	}
-
-	gvg, found := k.virtualGroupKeeper.GetGVG(ctx, lvg.GlobalVirtualGroupId)
-	if !found {
-		return fmt.Errorf("get GVG failed: %d, %s", lvg.GlobalVirtualGroupId, lvg.String())
-	}
-
-	secondaryRate := price.SecondaryStorePrice.MulInt(sdkmath.NewIntFromUint64(chargeSize)).TruncateInt()
-	secondaryRate = secondaryRate.MulRaw(int64(len(gvg.SecondarySpIds)))
-	if secondaryRate.IsPositive() {
-		secondaryTotalFlowRate = secondaryTotalFlowRate.Add(secondaryRate)
-		_, err = k.paymentKeeper.UpdateStreamRecordByAddr(ctx, types.NewDefaultStreamRecordChangeWithAddr(sdk.MustAccAddressFromHex(gvg.VirtualPaymentAddress)).
-			WithStaticBalanceChange(secondaryTotalFlowRate.MulRaw(timeToPay)))
-		if err != nil {
-			return fmt.Errorf("pay GVG failed: %s %s %s", bucketInfo.BucketName, objectInfo.ObjectName, err)
-		}
-	}
-
-	// validator tax rate
-	versionedParams, err := k.paymentKeeper.GetVersionedParamsWithTs(ctx, internalBucketInfo.PriceTime)
-	if err != nil {
-		return fmt.Errorf("failed to get validator tax rate: %w, time: %d", err, internalBucketInfo.PriceTime)
-	}
-	validatorTaxRate := versionedParams.ValidatorTaxRate.MulInt(primaryTotalFlowRate.Add(secondaryTotalFlowRate)).TruncateInt()
-	if validatorTaxRate.IsPositive() {
-		_, err = k.paymentKeeper.UpdateStreamRecordByAddr(ctx, types.NewDefaultStreamRecordChangeWithAddr(types.ValidatorTaxPoolAddress).
-			WithStaticBalanceChange(validatorTaxRate.MulRaw(timeToPay)))
+func (k Keeper) ChargeObjectStoreFeeForEarlyDeletion(ctx sdk.Context, userFlows []types.OutFlow, bucketInfo *storagetypes.BucketInfo, objectInfo *storagetypes.ObjectInfo, timeToPay int64) error {
+	totalStaticBalanceChange := sdkmath.NewInt(0)
+	for _, flow := range userFlows {
+		staticBalanceChange := flow.Rate.MulRaw(timeToPay)
```

Comment:
The merged outflow should be negative - because it is deletion.

Reply:
fixed